Generate protobuf files during compilation and fix the generation rules.

pivanof opened the pull request:
@enisoc @michael-berlin 

The change achieves the following:
1. Removes generated protobuf files for Go and Python and instead adds
   generation of these files during build by adding dependency of 'build' rule
   on 'proto' rule.
2. Modifies the generation of protobuf files to be more in line of how Makefiles
   should be written. This has an additional benefit that protobuf files won't
   be regenerated if proto files did not change.
3. Modifies the generation rules to work both locally on a workstation and
   inside the docker image. Docker image doesn't have /vt/dist/grpc installed,
   so the existing rule didn't work.

Note that as a result of this a new directory go/vt/.proto.tmp will exist after
building. I believe that should be fine and won't interfere with anything.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/youtube/vitess/1834)
<!-- Reviewable:end -->

enisoc commented:
For Go, one of the main reasons we check in the generated files is so that `go get` works for `vtctlclient`. We make use of this, for example, to greatly simplify the local setup required to run Vitess in Kubernetes: all you need locally is to have Go installed, and run the single `go get` (no bootstrap.sh, no dev.env required).

For Python, it may be workable but we should think about whether there are any downsides. For example, are there any situations (analogous to `go get`) where we currently are able to let users skip some bootstrap steps if all they want is to run Python client code against a remote Vitess cluster?

---

Review status: 0 of 50 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/1834#-:-KLY4VCur7hSmmkgOo7R:-570046727)*
<!-- Sent from Reviewable.io -->

pivanof commented:
Hm... Maybe instead of using `go get` we should have a script that downloads vitess/lite image and extracts vtctlclient from it? Though I guess that requires to have docker installed. Maybe we can just host pre-compiled vtctlclient somewhere? I believe github allows to have that.

---

Review status: 0 of 50 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/1834#-:-KLYAaiQhrKdEsanH6S8:462445653)*
<!-- Sent from Reviewable.io -->

enisoc commented:
For a Go project, `go get` is considered the gold standard of "easy installation" for a tool. I wouldn't want to abandon that compatibility for `vtctlclient` (the only thing you need if your cluster is remote) unless there is some really big benefit and no possible way to maintain `go get`-ability. In my opinion, having the proto files checked in is working reasonably well, so I don't see a big benefit to removing them. If you still disagree, we should discuss it more widely with the team.

---

Review status: 0 of 50 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/1834#-:-KLYnGONkHrm1XDHW5_j:1079895572)*
<!-- Sent from Reviewable.io -->

pivanof commented:
I don't like this, but I won't fight for this. I brought back the protobuf files, adjusted Makefile and .gitignore, and modified the desciption of the commit.

---

Review status: 0 of 2 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/1834#-:-KLZxU-95tvLrDLLwE3v:1957953880)*
<!-- Sent from Reviewable.io -->

michael-berlin commented:
Pavel, I feel your pain.

Last year I also talked with Alain about it and he convinced me that it was not worth fixing it. One later year I can live with the current situation. Also note that we have a "presubmit" check in Travis which verifies that the generated, checked in protobuf files are up to date: https://github.com/youtube/vitess/blob/master/.travis.yml#L106

Regarding this change: I like it. It looks cleaner this way. But I'm not a fan of the temporary directory and would prefer if we could avoid it.

---

Review status: 0 of 2 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[.gitignore, line 67 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1834#-KLZym5x6tfYtuJ-5yum:-KLZym5x6tfYtuJ-5yun:392316324) ([raw file](https://github.com/youtube/vitess/blob/c2a3c4280b7140b384d55c2243a0e7f6bd95213d/.gitignore#L67)):*
> ```
> 
> # generated protobuf files
> go/vt/.proto.tmp
> ```

Please add a leading `/` here. Otherwise, Git will match this path within every subtree.

---

*[Makefile, line 45 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1834#-KL_3EcbTthYqE_RdS95:-KL_3EcbTthYqE_RdS96:1526500284) ([raw file](https://github.com/youtube/vitess/blob/c2a3c4280b7140b384d55c2243a0e7f6bd95213d/Makefile#L45)):*
> ```Makefile
> clean:
> 	go clean -i ./go/...
> 	rm -rf third_party/acolyte go/vt/.proto.tmp
> ```

Please put this on a new line with a separate `rm` command because these two directories are not related.

---

*[Makefile, line 98 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1834#-KL_-GQOF0NmP_Hz5ej9:-KL_-GQPyOkXCWhn_xDU:1027779727) ([raw file](https://github.com/youtube/vitess/blob/c2a3c4280b7140b384d55c2243a0e7f6bd95213d/Makefile#L98)):*
> ```Makefile
> 
> PROTOC_DIR := $(VTROOT)/dist/grpc/usr/local/bin
> PROTOC_EXISTS := $(shell type -p $(PROTOC_DIR)/protoc)
> ```

Out of curiosity: What's the difference to a `test -x`?

Easier to use in the Makefile because it outputs nothing if the binary is not executable?

---

*[Makefile, line 100 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1834#-KL_-ruaad_UXtiy2IuA:-KL_-rubVPG2Ic8IsW5Y:689384849) ([raw file](https://github.com/youtube/vitess/blob/c2a3c4280b7140b384d55c2243a0e7f6bd95213d/Makefile#L100)):*
> ```Makefile
> PROTOC_EXISTS := $(shell type -p $(PROTOC_DIR)/protoc)
> ifeq (,$(PROTOC_EXISTS))
>   PROTOC_BINARY := $(shell which protoc)
> ```

This is a new feature.

I would rather not have it.

We regularly have problems with users who have an outdated protobuf version installed and that gets used somehow.

By always using the protoc from the gRPC installation which `bootstrap.sh` we reduce such a risk. In general, I don't think there's a usecase where you want to use a different `protoc` binary than the one we ship and compile for you.

---

*[Makefile, line 127 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1834#-KL_0d5NoTg-53DHuIO2:-KL_0d5NoTg-53DHuIO3:1079118675) ([raw file](https://github.com/youtube/vitess/blob/c2a3c4280b7140b384d55c2243a0e7f6bd95213d/Makefile#L127)):*
> ```Makefile
> 	for name in $(PROTO_SRC_NAMES); do \
> 	  mkdir -p go/vt/proto/$${name}; \
> 	  cp -a go/vt/.proto.tmp/$${name}.pb.go go/vt/proto/$${name}/$${name}.pb.go; \
> ```

This looks like the directory hierarchy between the generated files and what we use is changing?

Was this not necessary before because we did run protoc for .proto separately with a different output directory?

Personally, I would like to avoid that we need to copy files around because this way we can avoid the temporary directory.

---

*[Makefile, line 135 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1834#-KL_21pnr_IET2orZoWS:-KL_21pnr_IET2orZoWT:1680588034) ([raw file](https://github.com/youtube/vitess/blob/c2a3c4280b7140b384d55c2243a0e7f6bd95213d/Makefile#L135)):*
> ```Makefile
> 	mkdir -p go/vt/.proto.tmp
> 	$(PROTOC_DIR)/protoc -Iproto $< --go_out=plugins=grpc:go/vt/.proto.tmp
> 	sed --in-place -r -e 's,import ([a-z0-9_]+) ".",import \1 "github.com/youtube/vitess/go/vt/proto/\1",g' $@
> ```

It looks like we no longer need this fix. `protoc-gen-go` supports a parameter `import_prefix` which should take care of it:

https://github.com/golang/protobuf#parameters

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/1834#-:-KL_2XkYtAvqwECcDoli:-936725972)*
<!-- Sent from Reviewable.io -->

pivanof commented:
Review status: 0 of 2 files reviewed at latest revision, 6 unresolved discussions.

---

*[.gitignore, line 67 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1834#-KLZym5x6tfYtuJ-5yum:-KLb5i5U9_TiPSMZaG4q:-499146009) ([raw file](https://github.com/youtube/vitess/blob/c2a3c4280b7140b384d55c2243a0e7f6bd95213d/.gitignore#L67)):*
<details><summary><i>Previously, michael-berlin (Michael Berlin) wrote…</i></summary>
> Please add a leading `/` here. Otherwise, Git will match this path within every subtree.
</details>
Done.

---

*[Makefile, line 45 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1834#-KL_3EcbTthYqE_RdS95:-KLb5rVT2Cux4HEuPW3I:-499146009) ([raw file](https://github.com/youtube/vitess/blob/c2a3c4280b7140b384d55c2243a0e7f6bd95213d/Makefile#L45)):*
<details><summary><i>Previously, michael-berlin (Michael Berlin) wrote…</i></summary>
> Please put this on a new line with a separate `rm` command because these two directories are not related.
</details>
Done.

---

*[Makefile, line 98 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1834#-KL_-GQOF0NmP_Hz5ej9:-KLb5va_Jb_Y4Bqx1yln:-838385824) ([raw file](https://github.com/youtube/vitess/blob/c2a3c4280b7140b384d55c2243a0e7f6bd95213d/Makefile#L98)):*
<details><summary><i>Previously, michael-berlin (Michael Berlin) wrote…</i></summary>
> Out of curiosity: What's the difference to a `test -x`?
> 
> Easier to use in the Makefile because it outputs nothing if the binary is not executable?
</details>
The difference is that test -x returns result only by return code which is not easy to check in Makefile. type -p returns result by output which is very easy to use in an "ifeq".

---

*[Makefile, line 100 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1834#-KL_-ruaad_UXtiy2IuA:-KLb6FuXX6Z0UHbTuvgL:-1904840452) ([raw file](https://github.com/youtube/vitess/blob/c2a3c4280b7140b384d55c2243a0e7f6bd95213d/Makefile#L100)):*
<details><summary><i>Previously, michael-berlin (Michael Berlin) wrote…</i></summary>
> This is a new feature.
> 
> I would rather not have it.
> 
> We regularly have problems with users who have an outdated protobuf version installed and that gets used somehow.
> 
> By always using the protoc from the gRPC installation which `bootstrap.sh` we reduce such a risk. In general, I don't think there's a usecase where you want to use a different `protoc` binary than the one we ship and compile for you.
</details>
This is not a new feature, this is a bug fix. :)
There's no gRPC installation in the docker image. Or rather as Anthony said there is, but it is system-level. So without this it's impossible to run "make proto" in the docker image now. Are you suggesting that we should add gRPC to /vt/dist/grpc in the docker image?

---

*[Makefile, line 127 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1834#-KL_0d5NoTg-53DHuIO2:-KLb7M-5EIexiDad1Wzx:-681411120) ([raw file](https://github.com/youtube/vitess/blob/c2a3c4280b7140b384d55c2243a0e7f6bd95213d/Makefile#L127)):*
<details><summary><i>Previously, michael-berlin (Michael Berlin) wrote…</i></summary>
> This looks like the directory hierarchy between the generated files and what we use is changing?
> 
> Was this not necessary before because we did run protoc for .proto separately with a different output directory?
> 
> Personally, I would like to avoid that we need to copy files around because this way we can avoid the temporary directory.
</details>
No, the end result is not changing, location of $(PROTO_GO_OUTS) is the same.
This wasn't necessary before because we had fancy "find | sed | xargs" which was putting the name of the proto file into --go_out flag of protoc. So the name of the proto file ends up to appear twice in the path of the generated file (i.e. query.proto results in query/query.pb.go and vtrpc.proto results in vtrpc/vtrpc.pb.go). That's impossible to reflect in a per-file Makefile rule. That's why I do a per-file rule for $(PROTO_GO_TEMPS) which will convert e.g. proto/query.proto into go/vt/.proto.tmp/query.pb.go. And then I do one rule for all proto files which copies them from go/vt/.proto.tmp into the corresponding subdirectories of go/vt/proto.

If copy and .proto.tmp are no-go, then I'll have to resort back to the "find|sed|xargs" we have now, and then proto files will have to be compiled every time "make build" is run, even when proto files didn't change.

---

*[Makefile, line 135 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1834#-KL_21pnr_IET2orZoWS:-KLbDnBQGnA8feHI4GXI:667126153) ([raw file](https://github.com/youtube/vitess/blob/c2a3c4280b7140b384d55c2243a0e7f6bd95213d/Makefile#L135)):*
<details><summary><i>Previously, michael-berlin (Michael Berlin) wrote…</i></summary>
> It looks like we no longer need this fix. `protoc-gen-go` supports a parameter `import_prefix` which should take care of it:
> 
> https://github.com/golang/protobuf#parameters
</details>
I tried it and it doesn't work well. First of all it creates things like
`import proto "github.com/youtube/vitess/go/vt/proto/github.com/golang/protobuf/proto"`
And then instead of
`import query "github.com/youtube/vitess/go/vt/proto/query"`
it does
`import query "github.com/youtube/vitess/go/vt/proto/."`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/1834)*
<!-- Sent from Reviewable.io -->

pivanof commented:
Ping?
So is this a go or no-go?

---

Review status: 0 of 2 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/1834#-:-KM5LlIMC4ggvk9pMsjT:-1561979011)*
<!-- Sent from Reviewable.io -->

enisoc commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[Makefile, line 100 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1834#-KL_-ruaad_UXtiy2IuA:-KMRTjAss9hOPPHTabTr:-1770769970) ([raw file](https://github.com/youtube/vitess/blob/c2a3c4280b7140b384d55c2243a0e7f6bd95213d/Makefile#L100)):*
<details><summary><i>Previously, pivanof (Pavel Ivanov) wrote…</i></summary>
> This is not a new feature, this is a bug fix. :)
> There's no gRPC installation in the docker image. Or rather as Anthony said there is, but it is system-level. So without this it's impossible to run "make proto" in the docker image now. Are you suggesting that we should add gRPC to /vt/dist/grpc in the docker image?
</details>
If the local one exists, this will still prefer it, so that addresses the most common problem our users had. If the local one doesn't exist, either you never ran bootstrap (in which case everything will fail) or you ran it with --skip_root_installs - in which case you're explicitly telling us to use the system-installed stuff.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/1834#-:-KMRW7dXjT-IBzAVxuih:1417088721)*
<!-- Sent from Reviewable.io -->
